storage/src/main/java/io/delta/storage/HDFSLogStore.java

```diff
@@ -0,0 +1,176 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.storage;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.FileAlreadyExistsException;
+import java.util.EnumSet;
+import java.util.Iterator;
+import java.util.UUID;
+
+import io.delta.storage.internal.LogStoreErrors;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.*;
+import org.apache.hadoop.fs.CreateFlag;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link LogStore} implementation for HDFS, which uses Hadoop {@link FileContext} API's to
+ * provide the necessary atomic and durability guarantees:
+ * <ol>
+ *     <li>Atomic visibility of files: `FileContext.rename` is used write files which is atomic for HDFS.</li>
+ *     <li>Consistent file listing: HDFS file listing is consistent.</li>
+ * </ol>
+ */
+public class HDFSLogStore extends HadoopFileSystemLogStore {
+    private static final Logger LOG = LoggerFactory.getLogger(HDFSLogStore.class);
+    public static final String NO_ABSTRACT_FILE_SYSTEM_EXCEPTION_MESSAGE = "No AbstractFileSystem";
+
+    public HDFSLogStore(Configuration hadoopConf) {
+        super(hadoopConf);
+    }
+
+    @Override
+    public void write(
+            Path path,
+            Iterator<String> actions,
+            Boolean overwrite,
+            Configuration hadoopConf) throws IOException {
+        final boolean isLocalFs = path.getFileSystem(hadoopConf) instanceof RawLocalFileSystem;
+        if (isLocalFs) {
+            // We need to add `synchronized` for RawLocalFileSystem as its rename will not throw an
+            // exception when the target file exists. Hence we must make sure `exists + rename` in
+            // `writeInternal` for RawLocalFileSystem is atomic in our tests.
+            synchronized(this) {
+                writeInternal(path, actions, overwrite, hadoopConf);
+            }
+        } else {
+            // rename is atomic and also will fail when the target file exists. Not need to add the extra
+            // `synchronized`.
+            writeInternal(path, actions, overwrite, hadoopConf);
+        }
+    }
+
+    @Override
+    public Boolean isPartialWriteVisible(Path path, Configuration hadoopConf) {
+        return true;
+    }
+
+    /**
+     * @throws UnsupportedFileSystemException if the file system from the default configuration is
+     *                                        not supported
+     */
+    protected FileContext getFileContext(
+            Path path,
+            Configuration hadoopConf) throws UnsupportedFileSystemException {
+        return FileContext.getFileContext(path.toUri(), hadoopConf);
+    }
```

Comment:
nit: Could you inline this? It looks weird to have a new method that just calls another method.